rust/stub/src/main.rs

```diff
@@ -101,20 +95,37 @@ impl EmbeddedConfiguration {
 
 #[entry]
 fn main(handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
-    uefi_services::init(&mut system_table).unwrap();
+    // Initialize UEFI services (we need them for pretty much everything)
+    uefi_services::init(&mut system_table).expect("Failed to initialize UEFI services");
 
-    print_logo(system_table.stdout()).unwrap();
+    // Initialize logging before printing anything
+    uefi_log::init_from_cmdline(&mut system_table).expect("Failed to parse kernel command line");
+
+    print_logo();
+
+    let boot_services = system_table.boot_services();
 
     let config: EmbeddedConfiguration =
-        EmbeddedConfiguration::new(&mut booted_image_file(system_table.boot_services()).unwrap())
-            .expect("Failed to extract configuration from binary. Did you run lanzatool?");
+        match EmbeddedConfiguration::new(&mut match booted_image_file(boot_services) {
+            Ok(i) => i,
```

Comment:
These matches go away if you move all the code in main to a function that returns uefi::Result. Then you can use ? and map_err which is easier to read. The top level function can extract the status from the result via .status().